### ISSUE TITLE
Problem with implementation of optimizers.

### ISSUE THREAD

**Issue by SinghKislay** [mlpack/ensmallen#96]
So I have been looking for examples on how to use the different methods of FFN wrapper other than Add and Train no luck so far. So, I decided to give it a go on my understanding and so far it is going horrible.

I planned on implementing ResNet 
now all I wanted to do was

output=convolution(prev_input+prev_output)
but all the examples are straight forward model.Add and then optimize

So I planned to first add all the conv layer to the model by model.Add< xyzlayer<> >()
Then forward ops in a for loop followed by a loss function and  then optimizer.optimize

So here is a what I wrote:

//these are methods of ResNet Struct
mat  conv_forward(mat input){
         mat output;
         //We are trying to go just one convolution ahead;
	  model.Forward(input,output,iter,iter);
	  iter++;
          return output
  }
  

void  train(mat input, mat target){
          mat params =model.Parameters();
          error=this->loss_func(input, target);
          optimizer.optimize(Params,error);
          cout<<"iter :\n";
  }

double loss_func(mat input, mat target){
	  loss_val=loss.Forward(input,target);
    return loss_val;
  }

A little Insight would be very helpfull. 
ThankYou.

**Comment by rcurtin** [comment]
Hey there @SinghKislay, I think that this is not specific to ensmallen but instead mlpack.  In any case, I'd suggest taking a careful look at `src/mlpack/tests/feedforward_network_test.cpp` and `src/mlpack/tests/convolutional_network_test.cpp` and related files to get an idea of how to build neural networks with mlpack.  The strategy you described for building a network is pretty straightforward, but you don't need a special `ResNet` struct---you can just build an `FFN<>` and add the right layers.

**Comment by SinghKislay** [comment]
Hi, @rcurtin I have read the examples from test/convolutional_network_test.cpp, model/kaggle/DigitRecogniser/src/DigitRecogniser.cpp I have also read the example from vae in model reository. All of them use Add() method to add the layer and then they just use Train() method with input and target. I understad this and have also built one. But what I want to do is modify the outputs of each layer which is not done in any of the examples. Could u you point me to an example Where the Feedforward() method is used so that I could do f(x)+x. I don't know why is this so hard am I missing something here. Thankyou

**Comment by zoq** [comment]
@SinghKislay Sorry for the slow response, in this case you have to manually step through each layer or implement a new one that you could add right after each layer you like to modify. Here is the code to get the output after each layer:

https://github.com/mlpack/mlpack/blob/6faf0b782705c29267fbfeb1eca527b6006e5a04/src/mlpack/methods/ann/ffn_impl.hpp#L134-L146

in this case ``network[begin + i].OutputParameter()`` does contain the result. Let me know if I should clarify anything.

**Comment by SinghKislay** [comment]
@zoq thanks for reply, I understand we have to add layers manually, which is why I made the resnet struct on top of ffn class. Where I am facing problem is propagating the gradients back and the loss function. Say I push the input all the way through to the last layer and then push it into the loss fuction. How do I backprop. All the examples that I saw regarding back prop were using simple train method which we can't use in this case.
